### PR TITLE
Feature/password complexity

### PR DESCRIPTION
**Pull request by dariasamo**
## Description of changes

Password complexity options were added and validateInput function changed to validate password against complexity options. We can set some options to true by default, for example, asciiChar or digitChar to enforce some reasonable default, but user should be able to switch it off as well by setting option to false. If password doesn't meet some of the requirements, a message will be displayed with validation error detail.

## Related issues (if any)

https://github.com/keystonejs/keystone/issues/136

## Testing

- [x ] Please confirm `npm run test-all` ran successfully.

### CODE REVIEW

File: fields/types/password/PasswordType.js [hunk 1]
```diff
@@ -133,6 +150,17 @@ password.prototype.validateInput = function (data, callback) {
 		detail = 'passwords must match';
 	}
 	// TODO: we could support a password complexity option (or regexp) here
```
Comment (mxstbr) [inline fields/types/password/PasswordType.js]: This TODO is no longer needed I'm sure 😉 feel free to remove it!

File: fields/types/password/PasswordType.js [hunk 2]
```diff
@@ -133,6 +150,17 @@ password.prototype.validateInput = function (data, callback) {
...
+	if (min && typeof passwordValue === 'string') {
+		result = passwordValue.length > min;
+	}
+	for (var prop in complexity) {
+		if (complexity[prop] && typeof passwordValue === 'string') {
+			result = (regexChunk[prop]).test(passwordValue);
```
Comment (mxstbr) [inline fields/types/password/PasswordType.js]: There needs to be a check before here that makes sure that `regexChunk[prop]` exists, and if not console.logging the fact that the `complexity` option was configured wrongly on field with name `"whatever"`.

**Comment by mxstbr** [comment]
This is awesome, amazing job! :+1: 

I left two comments inline with small fixes, also there should be tests for each complexity option, validating both a positive and a negative result + the resulting error messages! Other than that this looks brilliant.

File: fields/types/password/PasswordType.js (hunk 2 above)
Comment (JedWatson) [inline fields/types/password/PasswordType.js]: We should be passing errors to the `callback`, not throwing

Comment (mxstbr) [inline fields/types/password/PasswordType.js]: This is a developer error, not a user error. The developer has a wrongly configured field…

Comment (JedWatson) [inline fields/types/password/PasswordType.js]: Oh right. If we're validating options, that should happen in the constructor and throw there (so it shows the error when keystone starts, not when validation is called while the app is running)

Comment (mxstbr) [inline fields/types/password/PasswordType.js]: Does that make sense @dariasamo?

Comment (dariasamo) [inline fields/types/password/PasswordType.js]: We think so ;) Let us amend PR with changes and we can always fix it if it's not correct.

File: fields/types/password/PasswordType.js [hunk 3]
```diff
@@ -123,16 +138,30 @@ password.prototype.compare = function (item, candidate, callback) {
  * Asynchronously confirms that the provided password is valid
  */
 password.prototype.validateInput = function (data, callback) {
-	var detail;
+	var detail = '';
```
Comment (mxstbr) [inline fields/types/password/PasswordType.js]: Can detail not be `undefined` here?

**Comment by mxstbr** [comment]
This looks awesome! One small question, what happens when you define a field with an invalid complexity option?

```JS
someField: {
  type: PasswordType,
  complexity: {
    doesntExist: true
  }
}
```

I think that _might_ throw an error?

**Comment by ostruzina** [comment]
@mxstbr: Yeah, good catch. This is actually still work in progress - we still need to work on the wrong config check. I didn't realize that my commit would be added into the PR automatically :)

**Comment by mxstbr** [comment]
No worries, it's good to keep the updates rolling 🚢

**Comment by mxstbr** [comment]
This is awesome, looks perfect! One tiny thing then this is ready to be merged, you'll need to guard the `for-in` statement where the validation of the complexity option happens. (see http://eslint.org/docs/rules/guard-for-in)

**Comment by mxstbr** [comment]
Awesome, amazing work @dariasamo @ostruzina!! 🎉

**Comment by webteckie** [comment]
This is great!   Can this PR be updated with Usage instructions so that when it is time to document this it will be trivial :-)  Or open a related issue in the doc repo and specify the usage there.

**Comment by dariasamo** [comment]
We will update PR with usage instructions then

**Comment by mxstbr** [comment]
> We will update PR with usage instructions then

No worries, submitting a PR to `keystonejs-site` with the documentation would be the best!

**Comment by ostruzina** [comment]
Ok! We plan to do that for all the new features we contributed with.

**Comment by webteckie** [comment]
Actually, isn't the Readme for the password field suppose to document the options?   They show pretty nice in fields explorer:-)
